### PR TITLE
Pull SE-0081 into Swift 3.0 Preview 1

### PR DESCRIPTION
**Pull request by tkremenek**
Pull SE-0081 into Swift 3.0 Preview 1.

### CODE REVIEW

**Comment by tkremenek** [comment]
@swift-ci smoke test OS X platform

**Comment by tkremenek** [comment]
@swift-ci test linux

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,32 @@ Note: This is in reverse chronological order, so newer entries are added to the
...
+    {
+        ...
+    }
+    ```
+
+  after:
+
+    ```swift
+    func anyCommonElements<T : SequenceType, U : SequenceType>(lhs: T, _ rhs: U) -> Bool
+        where T.Generator.Element: Equatable, T.Generator.Element == U.Generator.Element>
```
Comment (jckarter) [inline CHANGELOG.md]: Trailing `>` at the end of the line there.

**Comment by tkremenek** [comment]
@DougGregor can you review this?

**Comment by tkremenek** [comment]
@rjmccall can you review?

**Comment by rjmccall** [comment]
LGTM.

**Comment by DougGregor** [comment]
Note that this crashes if you add a where clause to something that isn't generic:

    protocol P { }

    struct X<T> {
      func f() where T : P { }
    }

**Comment by rjmccall** [comment]
Yeah, I realized after LGTM'ing that this case was problematic; sorry, Ted.  Doug and I talked about it, and the patch needs to be fixed to not make promises we're not ready for.

**Comment by tkremenek** [comment]
Reverted in #2860

**Comment by tkremenek** [comment]
@rjmccall @DougGregor do we have a bug report for this case?

**Comment by gribozavr** [comment]
@tkremenek We also need syntax highlighting, code completion support, formatting, and interface printing support.  It might just fall out from existing infrastructure, but it can also crash badly -- we don't know.  We need to add tests for these IDE features even if they happen to work perfectly fine with the new syntax.

**Pull request merged**